### ISSUE TITLE
Client does not gracefully handle large influx of transactions

### ISSUE THREAD

**Issue by nategraf** [celo-org/celo-blockchain#1169]
### Expected Behavior

When receiving a large influx of transactions, the node should continue to operate normally.
* Validators should continue to produce a block every 5 seconds
* Full nodes should continue to participate in the `p2p` protocol
* Nodes may discard transactions that it cannot process without violating it's uptime

### Current Behavior

On September 22, approximately 4000 transactions were sent to the Celo network. Block production stalled for 30 seconds, as the validators and proxies were overloaded, and 14 validator nodes experienced downtimes of greater than 1 minute. In these cases, the most common root cause was an OOM exception triggered by a spike in memory usage associated with processing the transactions.

An incident report is here https://docs.google.com/document/d/1ShECNeXiRp5-WDi6J4vxTFnFISZU94bKNHsibIwE4jQ/edit?usp=sharing
Additional analysis is here https://gist.github.com/nategraf/2fc2ef0e1a11da09da00cd7694c86da6#timeline-of-validator-recoveries

It is clear that the root cause of this issues is associated with the initial processing of transactions before they reach the transaction pool, but it is not clear what within the transaction intake process is consuming the most resources.

**Comment by hbandura** [comment]
Closing this issue with a likely explanation of what happened.

After reviewing the incident report, reading the source code for the peer2peer and transaction pool modules, and online discussion with @nategraf , I think the simplest explanation for what happened is the following:

The 4k transactions arrive at node A. The nodes adds them to its transaction pool, and forwards them to all of its peers P1…PN. They do the same, except they don’t send it to A since they know it already has the transactions. However, let’s say a node B is peer to M nodes of the P1…PN set. It will now receive M * 4k p2p messages at roughly the same time. This will occupy M * 4k * avg_transaction_size of memory instantly. Since M could easily have values near 100, it would not be strange to either provoke an OOM or degrade node performance in some way. Nodes crashing because of OOM triggered another known issue of corrupting on-disk chain data.

Hints that point to this explanation are:
- The p2p incoming messages buffer does not seem to have a message/size limit (there’s only a per-message limit)
- Nodes that first received the transactions didn’t have as many issues as the rest of the network (nodes more like node “A” in our example). I think it was also the case that proxies were affected but the nodes behind them weren’t as much (I didn’t see data that confirms this but if I remember correctly it’s what happened). Note that proxies have many peers and validators behind proxies do not.
- Our tx validation code is probably slower than ethereum’s, since we check for balance in non native currencies as well (e.g celo dollars).
- The thread that manages incoming messages blocks on the txpool.AddRemotes call, it doesn’t continue freeing messages until that call (that also validates all received txs) finishes. Therefore if for some reason the validation of a transaction (or transactions) takes a long time, the p2p blocks and won’t process any more messages (e.g NewBlockMsg, GetReceiptsMsg,  etc)
- The CELO version that was running at the time of the incident did not yet include ETH-65 (https://eips.ethereum.org/EIPS/eip-2464) ( == Celo66), which recognises the fact that tx messages make the bulk of all p2p traffic, and changes it to a notify-then-pull instead of just pushing the tx to the peers. It also ensures the txpool.AddRemotes call is in a different thread than the p2p incoming messages processing code.

I attempted to set up an environment for reproduction but I finally realised that it would take me more time than I had allotted for this. If someone with more familiarity with the private test environments can do so relatively quickly, it could be useful to see if this issue happens there. It does need to have at least one or two hundred full nodes, with a high number of peers each.

Since Celo66 solves the issue I believe is the cause for the incident, I believe that no extra work is needed on this front, other than maybe ensure that the txpool.AddRemotes is as fast as possible (which demands validation code to be fast too).

TLDR: p2p message explosion, no limit on incoming messages, only 1 thread for processing p2p messages, same thread emptying the p2p message queue validates txs = OOM and/or node stuck for a long time.
